enhancements/upstream-downstream-split.md

```diff
@@ -0,0 +1,228 @@
+
+### Implementation Details
+
+The downstream product is composed of the upstream projects plus downstream-only additions and modifications. Since downstream is expected to write its own entrypoints and extension packages, code from upstream dependencies is isolated from the accompanying downstream code.
+
+Downstream necessity still allows for direct modifications to copies of upstream dependencies, although ideally most problems can be resolved without doing so to reduce divergence from upstream. Backports will be supported for the previous [n-2] releases ([[n-5] cycles for releases with extended update support](https://access.redhat.com/support/policy/updates/openshift#ocp4_phases)).
+
+The `operator-framework/operator-lifecycle-manager`, `operator-framework/operator-registry` and `operator-framework/api` repositories will serve as the upstream repositories while a new repository `openshift/operator-lifecycle-manager` will be the downstream monorepo for all of the listed upstream repositories (and any that may be added in future).
+
+All downstream artifacts are built from a new downstream monorepo. Upstream operator-framework repositories are staged as git subtrees in a /staging directory, and corresponding “replace” directives exist in go.mod in order to resolve upstream dependencies from the copy in /staging, e.g.: “replace github.com/operator-framework/operator-lifecycle-manager => ./staging/operator-lifecycle-manager”.
+
+To minimize divergence we will have staging branches composed of upstream releases + a set of patches that are planned to be introduced upstream. If patches do not make it into the next upstream release, they will be cherry-picked into the next staging branch.
+
+### Downstreaming
+1. Change pushed to upstream/x/master
+2. Nightly job syncs upstream master with downstream using an [`UPSTREAM_MERGE.sh`](https://github.com/jmrodri/scripts/blob/master/UPSTREAM-MERGE.sh#L69) script for each subtree
+``` bash
+# UPSTREAM_MERGE.sh snippet
+if ! git diff-index HEAD --exit-code --quiet 2>&1 ; then
+	die "Working tree has modifications.  Cannot add."
+fi
+if ! git diff-index --cached HEAD --exit-code --quiet 2>&1 ; then
+	die "Index has modifications.  Cannot add."
+fi
+
+ref="${3:-master}"
+
+# check state of working directory
+git diff-index --quiet HEAD || { printf "!! Git status not clean, aborting !!\\n\\n%s" "$(git status)"; exit 1; }
+
+# update remote
+git fetch -t "$upstream_remote" "$ref"
+
+rev=$(git rev-parse $default --revs-only FETCH_HEAD) || exit $?
+
+if test "${#rev}" -ne 1; then
+	die "Multiple revisions found at HEAD: Expecting one, Got: '$rev'"
+fi
+
+git merge --no-commit -Xsubtree=$subtree_dir $rev
+
+# unmerged files are overwritten with the upstream copy
+unmerged_files=$(git diff --name-only --diff-filter=U --exit-code)
+differences=$?
+if [[ $differences -eq 1 ]]; then
+  unmerged_files_oneline=$(echo "$unmerged_files" | paste -s -d ' ')
+  git checkout --theirs -- $unmerged_files_oneline
```

Comment:
This will just move the conflicts out of a merge and into the application of each patch instead. This may be cleaner if we're only carrying a few patches. If we're failing on the merge here, it may be easier to use git subtree merge directly and avoid carrying patches altogether.